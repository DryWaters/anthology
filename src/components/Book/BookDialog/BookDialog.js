import React from "react";

import Button from "../../UI/Button/Button";

import classes from "./BookDialog.module.scss";

const BookDialog = props => (
  <div className={classes.BookDialog}>
    <div>
      <img
        className={classes.image}
        src={props.book.image}
        alt={props.book.title}
      />
    </div>
    <div>
      <p className={classes.prompt}>Are you sure you want to delete</p>{" "}
      <span className={classes.title}>{props.book.title}?</span>
    </div>
    <div>
      <Button btnType="danger">Delete</Button>
      <Button btnType="success">Cancel</Button>
    </div>
  </div>
);

export default BookDialog;
