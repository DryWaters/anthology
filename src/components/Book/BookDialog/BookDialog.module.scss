@import "../../../variables.scss";

.BookDialog {
  background-color: $color-light-brown;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-content: center;
  padding: 2rem;
  text-align: center;
}

.prompt {
  margin-top: 2rem;
}

.title {
  font-size: $text-medium-size;
  font-weight: bold;
}

@media (min-width: 500px) {
  .title {
    font-size: $text-large-size;
  }
}
