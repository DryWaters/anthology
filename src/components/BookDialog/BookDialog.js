import React from "react";
import PropTypes from "prop-types";

import BookImage from "../Book/BookImage/BookImage";
import Button from "../UI/Button/Button";

import classes from "./BookDialog.module.scss";

const BookDialog = props => {
  // Generate error message output if passed in from fetch()
  // Note: error message is cleared when modal is closed
  let errorMessage = null;
  if (props.errorMessage) {
    errorMessage = <p className="error">ERROR: {props.errorMessage}</p>;
  }

  return (
    <div className={classes.BookDialog}>
      <h2>Delete Book</h2>
      {errorMessage}
      <div>
        <BookImage title={props.book.title} image={props.book.image} />
      </div>
      <div>
        <p className={classes.prompt}>Are you sure you want to delete</p>
        <span className={classes.title}>{props.book.title}?</span>
      </div>
      <div>
        <Button btnType="cancel" clicked={props.onCancel}>
          Cancel
        </Button>
        <Button
          btnType="danger"
          clicked={() => props.onDelete(props.book.id, "DELETE")}
        >
          Delete
        </Button>
      </div>
    </div>
  );
};

BookDialog.propTypes = {
  book: PropTypes.shape({
    author: PropTypes.string.isRequired,
    description: PropTypes.string,
    id: PropTypes.string.isRequired,
    image: PropTypes.string,
    isbn: PropTypes.string,
    loaned: PropTypes.bool,
    title: PropTypes.string.isRequired
  }),
  onDelete: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  errorMessage: PropTypes.string
};

export default BookDialog;
