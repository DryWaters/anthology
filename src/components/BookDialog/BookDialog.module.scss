@import "../../variables.scss";

.BookDialog {
  align-content: center;
  background-color: $color-primary;
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 2rem;
  text-align: center;
}

.prompt {
  margin-top: 2rem;
}

.title {
  font-size: $text-medium-size;
  font-weight: bold;
}

@media (min-width: 500px) {
  .title {
    font-size: $text-large-size;
  }
}
