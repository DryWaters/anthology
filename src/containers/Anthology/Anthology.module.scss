@import "../../variables.scss";

.Anthology {
  padding-top: 10rem;
  background-color: $color-light-brown;
  min-height: 100vh;
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
}

@media (min-width: 500px) {
  .Anthology {
    padding-top: 20rem;
  }
}
