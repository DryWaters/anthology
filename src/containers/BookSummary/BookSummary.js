import React, { Component } from "react";

import uuidv4 from "uuid/v4";
import { validateInput, validateForm } from "../../shared/utility";

import Button from "../../components/UI/Button/Button";
import Input from "../../components/UI/Input/Input";
import classes from "./BookSummary.module.scss";

class BookSummary extends Component {
  state = {
    bookForm: {
      title: {
        elementType: "input",
        elementConfig: {
          type: "text",
          placeholder: "Book Title"
        },
        validation: {
          required: true
        },
        value: this.props.book ? this.props.book.title : "",
        touched: false
      },
      author: {
        elementType: "input",
        elementConfig: {
          type: "text",
          placeholder: "Author"
        },
        validation: {
          required: true
        },
        value: this.props.book ? this.props.book.author : "",
        touched: false
      },
      description: {
        elementType: "textarea",
        elementConfig: {
          type: "textarea",
          placeholder: "Description",
          rows: 5
        },
        value: this.props.book ? this.props.book.description : "",
        touched: false
      },
      image: {
        elementType: "input",
        elementConfig: {
          type: "text",
          placeholder: "Image URL"
        },
        validation: {
          required: true,
          validURL: true
        },
        value: this.props.book ? this.props.book.image : "",
        touched: false
      },
      loaned: {
        elementType: "input",
        elementConfig: {
          type: "checkbox"
        },
        checked: this.props.book ? this.props.book.loaned : false,
        touched: false
      }
    }
  };

  componentDidMount() {
    // validate form on passed in props
    let isFormValid = true;
    const newForm = { ...this.state.bookForm };
    for (let input in newForm) {
      newForm[input].valid = validateInput(newForm[input]);
      isFormValid = newForm[input].valid && isFormValid;
    }

    this.setState({
      bookForm: newForm,
      isFormValid
    });
  }

  inputChangedHandler = (event, input) => {
    const newForm = { ...this.state.bookForm };
    if (newForm[input].elementConfig.type === "checkbox") {
      newForm[input].checked = event.target.checked;
    } else {
      newForm[input].value = event.target.value;
    }
    newForm[input].valid = validateInput(newForm[input]);
    newForm[input].touched = true;

    const isFormValid = validateForm(newForm);

    this.setState({ bookForm: newForm, isFormValid });
  };

  formSubmitHandler = event => {
    event.preventDefault();
    if (!this.state.isFormValid) {
      return;
    }

    const { bookForm } = this.state;

    const newBook = {
      title: bookForm.title.value,
      author: bookForm.author.value,
      description: bookForm.description.value,
      image: bookForm.image.value,
      loaned: bookForm.loaned.checked
    };

    // updating a book
    if (this.props.book && this.props.book.id) {
      newBook.id = this.props.book.id;
      this.props.update(newBook, false);
    } else {
      // creating a new book
      newBook.id = uuidv4();
      this.props.update(newBook, true);
    }
  };

  render() {
    // convert obj of input fields to
    // array so that can be mapped over
    const formElementsArray = [];
    for (const key in this.state.bookForm) {
      formElementsArray.push({
        id: key,
        config: this.state.bookForm[key]
      });
    }
    let form = (
      <form onSubmit={this.formSubmitHandler}>
        {formElementsArray.map(formElement => (
          <Input
            key={formElement.id}
            label={formElement.id}
            elementType={formElement.config.elementType}
            elementConfig={formElement.config.elementConfig}
            checked={formElement.config.checked}
            value={formElement.config.value}
            invalid={!formElement.config.valid}
            touched={formElement.config.touched}
            changed={event => this.inputChangedHandler(event, formElement.id)}
          />
        ))}
        <Button btnType="danger" type="button" onCancel={this.props.cancel}>
          Cancel
        </Button>
        <Button
          type="submit"
          btnType="success"
          disabled={!this.state.isFormValid}
          onSubmit={this.formSubmitHandler}
        >
          {this.props.status}
        </Button>
      </form>
    );

    return (
      <div className={classes.BookSummary}>
        <h3>{this.props.status} Book</h3>
        <div>
          <img
            className={classes.image}
            src={
              this.props.book
                ? this.props.book.image
                : "./images/book-image-error.png"
            }
            onError={err => {
              err.target.onerror = null;
              err.target.src = "./images/book-image-error.png";
            }}
            alt="Book"
          />
        </div>
        {form}
      </div>
    );
  }
}

export default BookSummary;
