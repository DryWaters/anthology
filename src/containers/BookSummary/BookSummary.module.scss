@import "../../variables.scss";

.BookSummary {
  font-size: $text-medium-size;
  text-align: center;
  background-color: $color-primary;
}
