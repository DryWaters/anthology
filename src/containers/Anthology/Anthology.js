import React, { Component } from "react";

import Book from "../../components/Book/Book";
import BookSummary from "../../containers/BookSummary/BookSummary";
import Header from "../../components/UI/Header/Header";
import Modal from "../../components/UI/Modal/Modal";
import Spinner from "../../components/UI/Spinner/Spinner";

import classes from "./Anthology.module.scss";

class Anthology extends Component {
  state = {
    books: [],
    selectedBook: null,
    showSummary: false
  };

  componentDidMount() {
    fetch("http://localhost:5000/books")
      .then(response => response.json())
      .then(books => this.setState({ books }))
      .catch(err => console.log(err));
  }

  handleCloseSummary = () => {
    this.setState({
      showSummary: false
    });
  };

  handleDeleteBook = id => {};

  handleToggleLoan = id => {};

  handleSelectBook = id => {
    console.log("Clicked book");
  };

  render() {
    let books = <Spinner />;
    if (this.state.books) {
      books = this.state.books.map(book => (
        <Book
          key={book.id}
          clickBook={() => this.handleSelectBook(book.id)}
          deleteBook={() => this.handleDeleteBook(book.id)}
          toggleLoan={() => this.handleToggleLoan(book.id)}
          loaned={book.loaned}
          title={book.title}
          image={book.image}
        />
      ));
    }

    return (
      <div className={classes.Anthology}>
        <Header />
        <Modal
          show={this.state.showSummary}
          modalClose={this.handleCloseSummary}
        >
          <BookSummary />
        </Modal>
        {books}
      </div>
    );
  }
}

export default Anthology;
