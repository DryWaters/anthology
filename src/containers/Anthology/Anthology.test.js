import React from "react";
import { shallow } from "enzyme";

import Anthology from "./Anthology";
import BookSummary from "../BookSummary/BookSummary";
import BookDialog from "../../components/Book/BookDialog/BookDialog";

describe("<Anthology />", () => {
  it("should render", () => {
    const wrapper = shallow(<Anthology />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should return null if have no selected book", () => {
    const wrapper = shallow(<Anthology />);
    expect(wrapper.instance().getBookInformation()).toEqual(null);
  });

  it("should return book if have selected book", () => {
    const wrapper = shallow(<Anthology />);
    wrapper.setState({
      books: [
        {
          id: 1,
          title: "Something"
        },
        { id: 2, title: "Other book" }
      ],
      selectedBookId: 2
    });
    expect(wrapper.instance().getBookInformation()).toEqual({
      id: 2,
      title: "Other book"
    });
  });

  it("should return null if unable to match book", () => {
    const wrapper = shallow(<Anthology />);
    wrapper.setState({
      books: [
        {
          id: 1,
          title: "Something"
        },
        { id: 2, title: "Other book" }
      ],
      selectedBookId: 3
    });
    expect(wrapper.instance().getBookInformation()).toEqual(null);
  });

  it("should close modal and reset error state if call handleCloseModal", () => {
    const wrapper = shallow(<Anthology />);
    wrapper.setState({
      books: [],
      errorMessage: "Some error",
      modalContent: "Add",
      selectedBookId: 1
    });

    wrapper.instance().handleCloseModal();
    expect(wrapper.state()).toEqual({
      books: [],
      modalContent: null,
      errorMessage: null,
      selectedBookId: null
    });
  });

  it("should return a BookSummary if modalContent is Add/Update", () => {
    const wrapper = shallow(<Anthology />);
    wrapper.setState({
      books: [],
      errorMessage: "Some error",
      modalContent: "Add",
      selectedBookId: 1
    });
    wrapper.instance().displayModalContent();
    expect(wrapper.containsMatchingElement(<BookSummary />)).toEqual(true);
    wrapper.setState({
      modalContent: "Update"
    });
    wrapper.instance().displayModalContent();
    expect(wrapper.containsMatchingElement(<BookSummary />)).toEqual(true);
  });

  it("should return a BookDialog if modalContent is Delete", () => {
    const wrapper = shallow(<Anthology />);
    wrapper.setState({
      books: [],
      errorMessage: "Some error",
      modalContent: "Delete",
      selectedBookId: 1
    });
    wrapper.instance().displayModalContent();
    expect(wrapper.containsMatchingElement(<BookDialog />)).toEqual(true);
  });
});
