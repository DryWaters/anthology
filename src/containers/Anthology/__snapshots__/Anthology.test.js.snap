// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anthology /> should render 1`] = `ShallowWrapper {}`;
